loopout.go

```diff
@@ -1117,121 +1129,23 @@ func (s *loopOutSwap) waitForHtlcSpendConfirmed(globalCtx context.Context,
-				// way the server will know that the swap will
-				// go through and we'll be able to MuSig2
-				// cosign our sweep transaction. In the worst
-				// case if the server is uncooperative for any
-				// reason we can still sweep using scriptpath
-				// spend.
-				err = s.setStatePreimageRevealed(ctx)
-				if err != nil {
-					return nil, err
-				}
-
-				if !paymentComplete {
-					// Push the preimage for as long as the
-					// server is able to settle the swap
-					// invoice. So that we can continue
-					// with the MuSig2 sweep afterwards.
-					s.pushPreimage(ctx)
-				}
-
-				// Now attempt to publish a MuSig2 sweep txn.
-				// Only attempt at most maxMusigSweepRetires
-				// times to still leave time for an emergency
-				// script path sweep.
-				if musigSweepTryCount < maxMusigSweepRetries {
-					success := s.sweepMuSig2(
-						ctx, htlcOutpoint, htlcValue,
-					)
-					if !success {
-						musigSweepTryCount++
-					} else {
-						// Mark that we had a sweep
-						// that was successful. There's
-						// no need for the script spend
-						// now we can just keep pushing
-						// new sweeps to bump the fee.
-						musigSweepSuccess = true
-					}
-				} else if !musigSweepSuccess {
-					// Attempt to script path sweep. If the
-					// sweep fails, we can't do any better
-					// than go on and try again later as
-					// the preimage is alredy revealed and
-					// the server settled the swap payment.
-					// From the server's point of view the
-					// swap is succeeded at this point so
-					// we are free to retry as long as we
-					// want.
-					err := s.sweep(
-						ctx, htlcOutpoint, htlcValue,
-					)
-					if err != nil {
-						log.Warnf("Failed to publish "+
-							"non-cooperative "+
-							"sweep: %v", err)
-					}
-				}
-
-				// If the result of our spend func was that the
-				// swap has reached a final state, then we
-				// return nil spend details, because there is
-				// no further action required for this swap.
-				if s.state.Type() != loopdb.StateTypePending {
-					return nil, nil
-				}
-			} else {
-				err := s.sweep(ctx, htlcOutpoint, htlcValue)
-				if err != nil {
-					return nil, err
-				}
-
-				// If the result of our spend func was that the
-				// swap has reached a final state, then we
-				// return nil spend details, because there is no
-				// further action required for this swap.
-				if s.state.Type() != loopdb.StateTypePending {
-					return nil, nil
-				}
+			err = s.setStatePreimageRevealed(ctx)
+			if err != nil {
+				return nil, err
+			}
 
-				// If our off chain payment is not yet complete,
-				// we try to push our preimage to the server.
-				if !paymentComplete {
-					s.pushPreimage(ctx)
-				}
+			if !paymentComplete {
+				// Push the preimage for as long as the
+				// server is able to settle the swap
+				// invoice. So that we can continue
+				// with the MuSig2 sweep afterwards.
+				s.pushPreimage(ctx)
```

Comment:
This changes behavior slightly as for non-musig swaps we first make sure to publish the sweep (which also reveals the preimage) and then push it to the server.

Reply:
I think that by moving all sweeping logic  to the sweep batcher this information is no longer exposed (publish attempts). We are only ever notified when the outpoint is spent.

Reply:
We should just notify the swap about the successful publish.

Reply:
> We should just notify the swap about the successful publish.

that would lead to extra chans for communication between batcher<>swap?

Reply:
There's already the `sweepbatcher.SpendNotifier` which could notify if the sweep is published. Otherwise the client will just blindly reveal the preimage regardless. Not a huge deal if the swap is musig2 since that's default behavior but for legacy swaps the clients are less trusting.